Xamarin.Forms.Core/ResourceDictionary.cs

```diff
@@ -256,5 +285,66 @@ void OnValuesChanged(params KeyValuePair<string, object>[] values)
 		}
 
 		event EventHandler<ResourcesChangedEventArgs> ValuesChanged;
+
+		[Xaml.ProvideCompiled("Xamarin.Forms.Core.XamlC.RDSourceTypeConverter")]
+		public class RDSourceTypeConverter : TypeConverter, IExtendedTypeConverter
+		{
+			object IExtendedTypeConverter.ConvertFromInvariantString(string value, IServiceProvider serviceProvider)
+			{
+				if (serviceProvider == null)
+					throw new ArgumentNullException(nameof(serviceProvider));
+
+				var targetRD = (serviceProvider.GetService(typeof(Xaml.IProvideValueTarget)) as Xaml.IProvideValueTarget)?.TargetObject as ResourceDictionary;
+				if (targetRD == null)
+					return null;
+
+				var rootObjectType = (serviceProvider.GetService(typeof(Xaml.IRootObjectProvider)) as Xaml.IRootObjectProvider)?.RootObject.GetType().GetTypeInfo();
+				if (rootObjectType == null)
+					return null;
+
+				var lineInfo = (serviceProvider.GetService(typeof(Xaml.IXmlLineInfoProvider)) as Xaml.IXmlLineInfoProvider)?.XmlLineInfo;
+				var rootResourceId = rootObjectType.GetCustomAttribute<Xaml.XamlResourceIdAttribute>().ResourceId;
+				var rootNs = rootObjectType.GetCustomAttribute<Xaml.XamlResourceIdAttribute>().RootNamespace;
+				var uri = new Uri(value, UriKind.RelativeOrAbsolute);
+				var resourceId = ComputeResourceId(uri, rootResourceId, rootNs);
+				targetRD.SetAndLoadSource(uri, resourceId, rootObjectType.Assembly, lineInfo);
+				return uri;
+			}
+
+			internal static string ComputeResourceId(Uri uri, string rootResourceId, string rootNs)
+			{
+				if (uri.IsAbsoluteUri)
+					return $"{rootNs}{uri.AbsolutePath.Replace('/','.')}";
+
+				var rootResourceIdLength = rootResourceId.Length;
+				rootResourceIdLength = rootResourceId.LastIndexOf('.', rootResourceIdLength - 1); //drop the .xaml
```

Comment:
those length and index vars are used to avoid Substringing before the actual value is needed